### PR TITLE
[stable/kong] Options to install Prometheus plugin with ServiceMonitor

### PR DESCRIPTION
**Pull request by decayofmind**
#### What this PR does / why we need it:

This PR provides an option to install kong-prometheus plugin as KongPlugin custom resource.
Prometheus-operator ServiceMonitor could be installed as well.

As a side-effect of this PR, Kong CRDs are not being deleted with the release anymore. It prevents from losing Kong declarative configuration while uninstalling Kong.

The proposed change was inspired by Prometheus-operator and Velero charts.

#### Checklist

- [x] [DCO](https://github.com/helm/charts/blob/master/CONTRIBUTING.md#sign-your-work) signed
- [x] Chart Version bumped
- [x] Variables are documented in the README.md
- [x] Title of the PR starts with chart name (e.g. `[stable/chart]`)

### CODE REVIEW

**Comment by decayofmind** [comment]
@hbagdi any luck this will be useful?

**Comment by hbagdi** [comment]
@decayofmind Sorry for the delayed responses on this. Have been very busy with multiple things flying around. I'll try to get to this soon.

**Comment by hbagdi** [comment]
We will keep this PR around for now. 
There is a feature that needs to land in this Helm chart which exposes a metrics port on all Kong deployments, based on a ConfigMap, similar to how we do it [here](https://github.com/Kong/kubernetes-ingress-controller/blob/master/deploy/single/all-in-one-dbless.yaml).

Once that get's in, this PR can be simplified quite a bit:
- please do not over-complicate the logic of how installation is done for CRDs, let's keep it simple.
- While automatically enabling Prometheus plugin when ServiceMonitor is enabled, it is going a bit too far IMHO of providing a magical button for everything. It is okay to document thing and ask users to enable the plugin by themselves, depending on how they are running Kong (with or without Ingress Controller, with or without DB)

**Comment by hbagdi** [comment]
Since #15079 is merged in. We can work on this PR.
Please see my comment above.

Also, it probably might be more accurate to provide `serviceMonitor.enabled`, rather than `metrics.serviceMonitor.enabled` as a config value. Please refer to other stable charts in this repository.

**Comment by decayofmind** [comment]
@hbagdi `serviceMonitor.enabled` is under `metrics.`, because if `metrics` itself are disabled it has no sense. I can imagine a case, where someone will leave `metrics.enabled: false`, but `serviceMonitor.enabled: true`. 

But, if you think it's better, I have no objections to change the variable name as you said :)

**Comment by decayofmind** [comment]
> * please do not over-complicate the logic of how installation is done for CRDs, let's keep it simple.

For installation of CRDs I've used a built-in Helm feature/hook (https://github.com/helm/helm/blob/master/docs/charts_hooks.md#the-available-hooks) to install CRDs before any other template. If something else is over-complicated, let me now, I'll remove it from PR.

> * While automatically enabling Prometheus plugin when ServiceMonitor is enabled, it is going a bit too far IMHO of providing a magical button for everything. It is okay to document thing and ask users to enable the plugin by themselves, depending on how they are running Kong (with or without Ingress Controller, with or without DB)

A ServiceMonitor will be installed only if `metrics.serviceMonitor.create` set to `true`.

File: stable/kong/templates/servicemonitor.yaml [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+  namespace: {{ .Values.ingressController.metrics.serviceMonitor.namespace }}
+  {{- end }}
+  labels:
+    app: {{ template "kong.name" . }}
+    chart: "{{ .Chart.Name }}-{{ .Chart.Version }}"
+    heritage: {{ .Release.Service }}
+    release: {{ .Release.Name }}
+spec:
+  endpoints:
+  - port: kong-admin
```
Comment (hbagdi) [inline stable/kong/templates/servicemonitor.yaml]: With #15079 merged in, port 9542 will be used to export metrics. 
let's use that.
Also, it works only on HTTP scheme.

File: stable/kong/templates/cleanup-crds.yaml [hunk 2]
```diff
@@ -0,0 +1,40 @@
+{{- if .Values.ingressController.cleanUpCRDs }}
+apiVersion: batch/v1
```
Comment (hbagdi) [inline stable/kong/templates/cleanup-crds.yaml]: I've serious concerns of doing this as it might break existing deployments and managing definitions themselves in this way is way more riskier.
Let's remove this for now.

The `KongPlugin` created by Helm should be managed by Helm, meaning, Helm will delete it as part of this release.

File: stable/kong/templates/prometheus-plugin.yaml [hunk 3]
```diff
@@ -0,0 +1,13 @@
+{{- if and .Values.ingressController.enabled .Values.ingressController.metrics.enabled }}
+apiVersion: configuration.konghq.com/v1
+kind: KongPlugin
+metadata:
+  name: prometheus
```
Comment (hbagdi) [inline stable/kong/templates/prometheus-plugin.yaml]: Please include the release name in the name of this KongPlugin to avoid any collisions.

**Review by hbagdi — CHANGES_REQUESTED** [review]
Please rebase this PR on latest master.

File: stable/kong/templates/servicemonitor.yaml (hunk 1 above)
Comment (decayofmind) [reply stable/kong/templates/servicemonitor.yaml]: @hbagdi Does a custom-server-block, introduced in #15079 enables the Prometheus plugin? It seems like so.

Comment (hbagdi) [reply stable/kong/templates/servicemonitor.yaml]: Sort of.
It won't enable the plugin inside Kong. But it will expose the default metrics that are available even when the plugin is not enabled for any specific service.

File: stable/kong/templates/cleanup-crds.yaml (hunk 2 above)
Comment (decayofmind) [reply stable/kong/templates/cleanup-crds.yaml]: Agree

**Comment by decayofmind** [comment]
@hbagdi I've done some changes you requested, could you please take a look at my PR one more time.

File: stable/kong/templates/prometheus-plugin.yaml [hunk 4]
```diff
@@ -0,0 +1,13 @@
+{{- if and .Values.ingressController.enabled .Values.ingressController.metrics.enabled }}
```
Comment (hbagdi) [inline stable/kong/templates/prometheus-plugin.yaml]: As I said in a review before, let's remove this.
The user can create an additional KongPlugin resource if the need be after the Kong Ingress Controller is installed.

File: stable/kong/templates/deployment.yaml [hunk 5]
```diff
@@ -251,5 +251,5 @@ spec:
       volumes:
         - name: custom-nginx-template-volume
           configMap:
-            name: {{ .Release.Name }}-kong-default-custom-server-blocks
+            name: {{ template "kong.fullname" . }}-default-custom-server-blocks
```
Comment (hbagdi) [inline stable/kong/templates/deployment.yaml:267]: Why is this changed?

File: stable/kong/templates/crd-kongconsumer.yaml [hunk 6]
```diff
@@ -8,6 +8,9 @@ metadata:
     chart: "{{ .Chart.Name }}-{{ .Chart.Version }}"
     release: "{{ .Release.Name }}"
     heritage: "{{ .Release.Service }}"
+  annotations:
+    "helm.sh/hook": crd-install
+    "helm.sh/hook-delete-policy": "before-hook-creation"
```
Comment (hbagdi) [inline stable/kong/templates/crd-kongconsumer.yaml]: I believe these changes can go away once KongPlugin resource is removed.

File: stable/kong/templates/prometheus-plugin.yaml (hunk 4 above)
Comment (hbagdi) [reply stable/kong/templates/prometheus-plugin.yaml]: Do note that Kong will expose some basic observability stats even if this plugin is not enabled.

Comment (decayofmind) [reply stable/kong/templates/prometheus-plugin.yaml]: OK, let it be as you're suggesting, I've got the point :)

File: stable/kong/templates/servicemonitor.yaml (hunk 1 above)
Comment (decayofmind) [reply stable/kong/templates/servicemonitor.yaml]: Right now, with the `custom-server-block` present by default, this PR will only optionally add a `ServiceMonitor`

File: stable/kong/templates/deployment.yaml (hunk 5 above)
Comment (decayofmind) [reply stable/kong/templates/deployment.yaml:267]: It may not be the matter of this PR, but in general, all objects in a chart should be addressing `{{ template "<chartname.fullname>" }}` in its names.

It allows a user to override objects naming for the whole installation, by taking into account `.Values.nameOverride` and `.Values.fullNameOverride` (not yet in Kong chart) variables passed to the chart.

File: stable/kong/templates/crd-kongplugins.yaml [hunk 7]
```diff
@@ -8,6 +8,9 @@ metadata:
     chart: "{{ .Chart.Name }}-{{ .Chart.Version }}"
     release: "{{ .Release.Name }}"
     heritage: "{{ .Release.Service }}"
+  annotations:
+    "helm.sh/hook": crd-install
+    "helm.sh/hook-delete-policy": "before-hook-creation"
```
Comment (hbagdi) [inline stable/kong/templates/crd-kongplugins.yaml]: I think these changes to CRDs can be removed since the Prometheus plugin is not being automatically created.

Another problem that I see here:
If multiple Kong Ingress Controllers are being installed one after the other, then this will incorrectly delete the CRD first, and then re-install the CRD. This will lead to deletion of all the custom resources that were present in the cluster before the installation.

**Comment by hbagdi** [comment]
@decayofmind Left one comment and other than that, this PR looks good.
#16293 will be merged before this so please rebase again.

**Comment by hbagdi** [comment]
Please rebase on master. This is real close and almost ready to merge.

File: stable/kong/templates/crd-kongplugins.yaml (hunk 7 above)
Comment (hbagdi) [reply stable/kong/templates/crd-kongplugins.yaml]: @decayofmind Could you address this comment?
Thanks!

**Comment by hbagdi** [comment]
@decayofmind Please address the one outstanding comment and fix the lint failure.

Once that's done, we are good to merge.

**Comment by decayofmind** [comment]
@hbagdi It seems like everything is finally fixed

**Comment by hbagdi** [comment]
@decayofmind Thank you for your contribution!
